core/src/main/java/org/apache/iceberg/avro/ApplyNameMapping.java

```diff
@@ -0,0 +1,182 @@
+import java.util.stream.Collectors;
+import org.apache.avro.LogicalType;
+import org.apache.avro.Schema;
+import org.apache.iceberg.mapping.MappedField;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/**
+ * An Avro Schema visitor to apply a name mapping to add Iceberg field IDs.
+ *
+ * <p>Methods return null when a schema has no ID and cannot be projected.
+ */
+public class ApplyNameMapping extends AvroSchemaVisitor<Schema> {
+  private final NameMapping nameMapping;
+
+  public ApplyNameMapping(NameMapping nameMapping) {
+    this.nameMapping = nameMapping;
+  }
+
+  @Override
+  public Schema record(Schema record, List<String> names, List<Schema> fields) {
+    List<Schema.Field> originalFields = record.getFields();
+
+    List<Schema.Field> newFields = Lists.newArrayList();
+    for (int i = 0; i < originalFields.size(); i += 1) {
+      Schema newSchema = fields.get(i);
+      if (newSchema != null) {
+        Schema.Field field = originalFields.get(i);
+        Integer fieldId = AvroSchemaUtil.getFieldId(field, nameMapping, fieldNames());
+        if (fieldId != null) {
+          // always copy because fields can't be reused
+          newFields.add(copyField(field, newSchema, fieldId));
+        }
+      }
+    }
+
+    return copyRecord(record, newFields);
+  }
+
+  @Override
+  public Schema union(Schema union, List<Schema> options) {
+    if (options.equals(union.getTypes())) {
+      return union;
+    }
+
+    List<Schema> validOptions =
+        options.stream().filter(Objects::nonNull).collect(Collectors.toList());
+
+    return copyProps(union, Schema.createUnion(validOptions));
+  }
+
+  @Override
+  public Schema array(Schema array, Schema element) {
+    if (array.getLogicalType() instanceof LogicalMap
+        || (isKeyValueMapping(fieldNames())
+            && AvroSchemaUtil.isKeyValueSchema(array.getElementType()))) {
+      return copyProps(array, Schema.createArray(element));
+    }
+
+    Integer elementId = AvroSchemaUtil.elementId(array);
+    if (elementId != null) {
+      if (array.getElementType().equals(element)) {
+        return array;
+      }
+
+      return copyProps(array, createArray(element, elementId));
+    }
+
+    MappedField mapping = nameMapping.find(fieldNames(), "element");
+    if (mapping != null) {
+      return copyProps(array, createArray(element, mapping.id()));
+    }
+
+    return null;
+  }
+
+  private boolean isKeyValueMapping(Iterable<String> names) {
+    return nameMapping.find(names, "key") != null && nameMapping.find(names, "value") != null;
+  }
+
+  @Override
+  public Schema map(Schema map, Schema value) {
+    Integer keyId = AvroSchemaUtil.keyId(map);
+    Integer valueId = AvroSchemaUtil.valueId(map);
+    if (keyId != null && valueId != null) {
+      if (map.getValueType().equals(value)) {
+        return map;
+      }
+
+      return copyProps(map, createMap(value, keyId, valueId));
+    }
```

Comment:
I prefer not to do multiple map lookups if it's avoidable. I don't think that it is too bad to have the methods, especially since I'm using them in later changes.